File: packages/agent-sdk-python/src/agentmarket_sdk/http.py
"""HTTP helpers — the Python port of `packages/agent-sdk/src/http.ts`."""

from __future__ import annotations

from typing import NoReturn
from urllib.parse import urlencode

import httpx

from .errors import HttpError, PaymentFailedError
from .types import CallParams, PaymentRequiredResponse

# Codes the payment gate itself raises (see apps/api/src/middleware/x402-payment.ts)
# — as distinct from a plain validation/not-found/rate-limit failure.
PAYMENT_ERROR_CODES = frozenset(
    {"PAYMENT_INVALID", "PAYMENT_VERIFICATION_FAILED", "PAYMENT_ALREADY_SETTLED", "BUDGET_EXCEEDED"}
)


def build_url(base_url: str, resource: str, params: CallParams | None = None) -> str:
    base = resource if resource.startswith("http") else f"{base_url}{'' if resource.startswith('/') else '/'}{resource}"
    if not params:
        return base
    query = urlencode({k: v for k, v in params.items() if v is not None})
    if not query:
        return base
    separator = "&" if "?" in base else "?"
    return f"{base}{separator}{query}"


def read_error(res: httpx.Response) -> NoReturn:
    """Reads a non-2xx, non-402 response body and raises the appropriately-typed error. Always raises — the `NoReturn` type mirrors the TS SDK's `never` so callers can `return read_error(res)`."""
    try:
        body = res.json()
    except ValueError:
        body = {}

    error = body.get("error", {}) if isinstance(body, dict) else {}
    code = error.get("code", "UNKNOWN_ERROR")
    message = error.get("message", f"Request failed with HTTP {res.status_code}")

    if code in PAYMENT_ERROR_CODES:
        raise PaymentFailedError(code, message, error.get("details"))
    raise HttpError(res.status_code, code, message, error.get("details"))


def read_payment_required(res: httpx.Response) -> PaymentRequiredResponse:
    return PaymentRequiredResponse.from_json(res.json())

File: packages/agent-sdk/src/http.ts
import type { PaymentRequiredResponse } from '@rohankumar4179/shared-types';
import { HttpError, PaymentFailedError } from './errors.js';
import type { CallParams } from './types.js';

export function buildUrl(baseUrl: string, resource: string, params?: CallParams): string {
  const base = resource.startsWith('http') ? resource : `${baseUrl}${resource.startsWith('/') ? '' : '/'}${resource}`;
  const url = new URL(base);
  if (params) {
    for (const [key, value] of Object.entries(params)) {
      if (value !== undefined) url.searchParams.set(key, String(value));
    }
  }
  return url.toString();
}

interface RawErrorBody {
  error?: { code?: string; message?: string; details?: Record<string, unknown> };
}

/** Codes the payment gate itself raises (see apps/api/src/middleware/x402-payment.ts) — as distinct from a plain validation/not-found/rate-limit failure. */
const PAYMENT_ERROR_CODES = new Set([
  'PAYMENT_INVALID',
  'PAYMENT_VERIFICATION_FAILED',
  'PAYMENT_ALREADY_SETTLED',
  'BUDGET_EXCEEDED',
]);

/** Reads a non-2xx, non-402 response body and throws the appropriately-typed error. Always throws — return type is `never` so callers can `return readError(res)` from a function that otherwise returns a value. */
export async function readError(res: Response): Promise<never> {
  const body = (await res.json().catch(() => ({}))) as RawErrorBody;
  const code = body.error?.code ?? 'UNKNOWN_ERROR';
  const message = body.error?.message ?? `Request failed with HTTP ${res.status}`;

  if (PAYMENT_ERROR_CODES.has(code)) {
    throw new PaymentFailedError(code, message, body.error?.details);
  }
  throw new HttpError(res.status, code, message, body.error?.details);
}

export async function readPaymentRequired(res: Response): Promise<PaymentRequiredResponse> {
  return (await res.json()) as PaymentRequiredResponse;
}
